### ISSUE TITLE
Custom Xpath jar is causing while building EAR using mvn plugin

### ISSUE THREAD

**Issue by thiyagu19841** [TIBCOSoftware/bw6-plugin-maven#176]
I have custom xpath jar and sharedModule jar dependency from application module (TIBCO BCE)..  Maven plugin 1.3.0 is not able to generate ear if I keep custom xpath jar in POM.xml dependency section and getting below error.. However, ear is getting generated if I remove this custom xpath jar from dependency section..  Can you please help me on this and what is causing below null pointer exception..

[INFO] Updated the Manifest version 
[INFO] The OSGi verion is 2.0.10 for Maven version of 2.0.10
[INFO] Adding Modules to the EAR file
[INFO] Adding Module JAR with name orders-flow-events_2.0.10.jar with version 2.0.10
[INFO] Adding Module JAR with name ErrorandLogging-2.0.6.jar with version 2.0.6
[INFO] Adding Module JAR with name ErrorandLogging_CustomXPathFunctions-2.0.2.jar with version 2.0.2
[ERROR] Failed to add modules to the Application
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] orders-flow-events ................................. SUCCESS [01:32 min]
[INFO] orders-flow-events.application ..................... FAILURE [  7.654 s]
[INFO] orders-flow-events.application.parent .............. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:07 min
[INFO] Finished at: 2018-08-20T22:56:49+00:00
[INFO] Final Memory: 63M/1066M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.tibco.plugins:bw6-maven-plugin:1.3.0.002:bwear (default-bwear) on project orders-flow-events.application: Failed to create BW EAR Archive: NullPointerException -> [Help 1]

**Comment by sonalshingavi** [comment]
Closing the issue, Please re-open if this still has a issue with 2.0.1.

**Comment by thiyagu19841** [comment]
The issue is that CustomXpath plugin project has dependency on below jar in POM.xml; though our internal maven repo has all the plugin jars,  below jar is not getting added while building jar for this CustomXpath plugin project via tibco mvn plugin; hence it is causing null pointer exception while building ear for application module which has dependency on this CustomXpath plugin project..  In order to resolve this issue, I just kept this jar (com.tibco.xml.cxf.common.13.200.jar) in lib folder explicitly from CustomXpath plug-in project; ran mvn command to build jar and then EAR is able to built successfully..   However, can you check why tibco maven plugin is not able to download this dependency jars from internal repository while generation jar for CustomXpath plugin-in project. 
 
<dependency>
        <groupId>com.tibco.plugins</groupId>
        <artifactId>com.tibco.xml.cxf.common</artifactId>
        <version>1.3.200</version>
 </dependency>

**Comment by thiyagu19841** [comment]
The issue is that CustomXpath plugin project has dependency on below jar in POM.xml; though our internal maven repo has all the plugin jars,  below jar is not getting added while building jar for this CustomXpath plugin project via tibco mvn plugin; hence it is causing null pointer exception while building ear for application module which has dependency on this CustomXpath plugin project..  In order to resolve this issue, I just kept this jar (com.tibco.xml.cxf.common.13.200.jar) in lib folder explicitly from CustomXpath plug-in project; ran mvn command to build jar and then EAR is able to built successfully..   However, can you check why tibco maven plugin is not able to download this dependency jars from internal repository while generation jar for CustomXpath plugin-in project.  Please reopen this issue
 
<dependency>
        <groupId>com.tibco.plugins</groupId>
        <artifactId>com.tibco.xml.cxf.common</artifactId>
        <version>1.3.200</version>
 </dependency>

**Comment by SujataDarekar** [comment]
@thiyagu19841  Basically Maven search for dependency either in local .m2 repository or maven central repository but if the dependency JARS are present in your internal maven repo then we need to add these internal repository in pom.xml under repositories section. So that maven plugin will search for dependency in your internal repo as well.

**Comment by thiyagu19841** [comment]
@SujataDarekar Yes.. That's correct..  Though I mentioned our internal repo in pom.xml and dependent jars are available in our internal repo, tibco maven plugin is not able to build custom xpath jar with this dependent jar.

**Comment by SujataDarekar** [comment]
@thiyagu19841  Could you please share you project?

**Comment by SujataDarekar** [comment]
@thiyagu19841 
I have tried to reproduce the issue by following Steps:
Scenario 1:
1. Created internal maven repository and installed the customXpath Jar on my maven internal repository.
2. Added the credential details of internal repository in maven setting file.
3. In my BW project, I have added the customXpath jar dependacy through POM file by giving the groupId, ArtifactID,Version
4. Also in the POM file, I have added the repository details in the repositories section.
5. Ran the "package" goal on parent project.

Scenario 2:
Added the com.tibco.xml.cxf.common jar on internal maven repository and then in CustomXpath provided the dependeny of the same JAR.

In both scenarios EAR get generated successfully and I am not able to reproduce the issue.
So basically maven plugin adds the dependacy JARs  which is present on internal maven repository also.
Please let me know if I need to follow the another steps to reproduce the issue

**Comment by thiyagu19841** [comment]
Hi,

Do you see com.tibco.xml.cxf.common.jar is available in lib folder from customXpath.jar.. Can you please share CustomXpath.jar.

Thanks and regards,
Thiyagarajan K.

**Comment by SujataDarekar** [comment]
Hi @thiyagu19841 
com.tibco.xml.cxf.common.jar does not  get included in lib folder because in maven plugin we have validation to not include JAR com.tibco.xml.cxf.common.jar in lib folder.
The reason is that these JAR is required only at compile time and at runtime this will be provided by Tibco environment.

**Comment by SujataDarekar** [comment]
@thiyagu19841 Please let me know if you have any another issue or any concerns.

**Comment by SujataDarekar** [comment]
As discussed If Custom-XPath Plugin having a dependency on JARS present at Tibco-home then we are providing resolution to add those dependencies as a "System" scope and provide dependencies path in the System path in the future release. Also, the scenario that Maven Plugin unable to download dependency from the maven internal repository is not able to reproduce. In fact, Maven Plugin downloads the dependencies from an internal maven repository as well. Only we need to provide the entry of repository in POM.xml under the repositories section. Hence closing the issue.
